src/components/app.rs

```diff
@@ -8,47 +9,127 @@ use super::{
     users_cache::{self, UsersCacheComponent},
 };
 
+use std::sync::Mutex;
+
 pub struct AppComponents {
     pub health: HealthComponent,
     pub synapse: SynapseComponent,
     pub config: Config,
     pub db: DatabaseComponent,
-    pub user_cache: UsersCacheComponent,
+    pub users_cache: Mutex<UsersCacheComponent>,
 }
 
-impl AppComponents {
-    pub async fn new(custom_config: Option<Config>) -> Self {
-        if let Err(_) = env_logger::try_init() {
-            log::debug!("Logger already init")
-        }
+pub struct CustomComponents {
+    pub synapse: Option<SynapseComponent>,
+    pub db: Option<DatabaseComponent>,
+    pub users_cache: Option<UsersCacheComponent>,
+    pub redis: Option<Redis>,
+}
 
-        // Initialize components
+impl AppComponents {
+    pub async fn new(
+        custom_config: Option<Config>,
+        custom_components: Option<CustomComponents>,
+    ) -> Self {
         let config = custom_config
             .unwrap_or_else(|| Config::new().expect("Couldn't read the configuration"));
+        if custom_components.is_none() {
+            AppComponents::default(config).await
+        } else {
+            // For testing propouses mainly
```

Comment:
Spelling for US English:
```suggestion
            // For testing proposes mainly
```